src/server/array_func.c

```diff
@@ -628,57 +643,27 @@ static struct ajtrkhd *mk_subjob_index_tbl(char *range, int initalstate)
 int
 setup_arrayjob_attrs(attribute *pattr, void *pobj, int mode)
 {
-	int  i;
 	job *pjob = pobj;
 
-	if (mode == ATR_ACTION_NEW) {
-		/* validate max array size */
-		int l, x, y, z, ct;
-		char *ep;
-
-		if (server.sv_attr[(int)SVR_ATR_maxarraysize].at_flags & ATR_VFLAG_SET)
-			l = server.sv_attr[(int)SVR_ATR_maxarraysize].at_val.at_long;
-		else
-			l = PBS_MAX_ARRAY_JOB_DFL;  /* default limit 10000 */
-		if (parse_subjob_index(pattr->at_val.at_str, &ep, &x, &y, &z, &ct) != 0)
-			return PBSE_BADATVAL;
-		if (ct > l)
-			return PBSE_MaxArraySize;
-
-	}
-
 	/* set attribute "array" True  and clear "array_state_count" */
 
 	pjob->ji_wattr[(int)JOB_ATR_array].at_val.at_long = 1;
 	pjob->ji_wattr[(int)JOB_ATR_array].at_flags = ATR_VFLAG_SET | ATR_VFLAG_MODCACHE;
 	job_attr_def[(int)JOB_ATR_array_state_count].at_free(&pjob->ji_wattr[(int)JOB_ATR_array_state_count]);
 
-	if (mode == ATR_ACTION_RECOV) {
-		int x, y, z, ct;
-		char *ep;
-
-		/* on recovery ... */
-		/* parse the various components again, since we dont store them */
-		if (parse_subjob_index(pattr->at_val.at_str, &ep, &x, &y, &z, &ct) != 0)
-			return PBSE_BADATVAL;
-
-		pjob->ji_ajtrk->tkm_ct = ct;
-		pjob->ji_ajtrk->tkm_step = z;
-		pjob->ji_ajtrk->tkm_flags = 0;
-
-		/* reset counts and any running/exiting subjob to queued */
-		for (i=0; i < PBS_NUMJOBSTATE; ++i)
-			pjob->ji_ajtrk->tkm_subjsct[i] = 0;
-		for (i=0; i < pjob->ji_ajtrk->tkm_ct; ++i) {
-			if ((pjob->ji_ajtrk->tkm_tbl[i].trk_status == JOB_STATE_RUNNING)
```

Comment:
You seem to have correctly eliminated the need to use the tkm_tbl[n] array field. However, have we removed the definition from the job.h structure?

Reply:
I agree we are having a duplication of data in memory and the array in the tracking table should be unnecessary.

Regarding the estimation of effort needed: Seeing in the code that tkm_tbl is used more than forty places and seems to be tightly woven around. Removing it with proper quality can easily be a man month effort.